### PR TITLE
Workshop and profile update requests

### LINKED ISSUE #457: integrate workshop requests in to Amy
I was just talking to Giacomo and sharing our workshop request form with him, and an idea came up. This might be beyond the scope of Amy, but I wonder if we can integrate workshop requests in to Amy. 

The idea would be that there would be a front facing form for a workshop request, then instead of the information going in to another database or Google doc, that info would go in to Amy. That would let us better consolidate workshop coordination, and get workshops in to the system right at the start. Then we could just add the instructors and URLs as we went along.

### PR DESCRIPTION
**Pull request by pbanaszkiewicz**
Fixes #457 

This is just the beginning of the user interface. Admin interface is missing, and so are other functionalities.

### CODE REVIEW

**Comment by pbanaszkiewicz** [comment]
All right, this one is almost finished. I need to add tests.

I'll postpone the deadline on v0.9 by a few days, since @gvwilson will be available to review no sooner than 30.08.

**Comment by pbanaszkiewicz** [comment]
This PR is quite big. Here's a short list of what's left to do:

* update code to latest permissions PR #472 
* upgrade workshop request to Data Carpentry's form
* add a selectable country field in that form
* bring back the "Import from URL" for new events done via workshop request
* set up email (**important**)
* add tests

File: amy/settings.py [hunk 1]
```diff
@@ -32,6 +32,51 @@
...
+    assert RECAPTCHA_PUBLIC_KEY, 'RECAPTCHA site key not present'
+    assert RECAPTCHA_PRIVATE_KEY, 'RECAPTCHA secure key not present'
+
+# email settings
+ADMINS = (
+    ('Piotr Banaszkiewicz', 'piotr@banaszkiewicz.org'),
+    ('Greg Wilson', 'gvwilson@software-carpentry.org'),
+)
+# "From:" for error messages sent out to ADMINS
+SERVER_EMAIL = os.environ.get('AMY_SERVER_EMAIL', 'root@localhost')
```
Comment (gvwilson) [inline amy/settings.py]: Is `root@localhost` a useful default?

File: amy/settings.py [hunk 2]
```diff
@@ -32,6 +32,51 @@
...
+
+# addresses to receive "New workshop request" or "New profile update request"
+# notifications
+REQUEST_NOTIFICATIONS_RECIPIENTS = (
+    'admins-all@software-carpentry.org',
```
Comment (gvwilson) [inline amy/settings.py]: I think it's `admin-all@software-carpentry.org` (without the 's')

File: amy/settings.py [hunk 3]
```diff
@@ -32,6 +32,51 @@
...
+)
+EMAIL_HOST = os.environ.get('AMY_EMAIL_HOST', 'localhost')
+EMAIL_HOST_USER = os.environ.get('AMY_EMAIL_HOST_USER', '')
+EMAIL_HOST_PASSWORD = os.environ.get('AMY_EMAIL_HOST_PASSWORD', '')
+EMAIL_PORT = int(os.environ.get('AMY_EMAIL_PORT', 25))
+EMAIL_TIMEOUT = 10  # timeout for blocking email operations, in seconds
```
Comment (gvwilson) [inline amy/settings.py]: That's a lot of environment variables...

File: workshops/forms.py [hunk 4]
```diff
@@ -37,6 +39,10 @@ def __init__(self, form=None):
         self.inputs.append(Submit('submit', 'Submit'))
 
 
+class BootstrapHelperGet(BootstrapHelper):
```
Comment (gvwilson) [inline workshops/forms.py]: docstring?

File: workshops/forms.py [hunk 5]
```diff
@@ -429,3 +436,71 @@ class HostForm(forms.ModelForm):
     class Meta:
         model = Host
         fields = ['domain', 'fullname', 'country', 'notes']
+
+
+class SwCEventRequestForm(forms.ModelForm):
```
Comment (gvwilson) [inline workshops/forms.py]: docstring

File: workshops/forms.py [hunk 6]
```diff
@@ -429,3 +436,71 @@ class HostForm(forms.ModelForm):
...
+            'admin_fee_payment': forms.RadioSelect(),
+        }
+
+
+class DCEventRequestForm(SwCEventRequestForm):
+    workshop_type = forms.CharField(initial='dc', widget=forms.HiddenInput())
+    understand_admin_fee = forms.BooleanField(
+        initial=False,
+        label='I understand the Data Carpentry\'s administration fee.',
+        help_text='There is a per-workshop fee for Data Carpentry to cover '
```
Comment (gvwilson) [inline workshops/forms.py]: We should probably have this for Software Carpentry workshops as well - please check wording with Maneesha.

File: workshops/migrations/0040_auto_20150827_1715.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+def add_academic_levels(apps, schema_editor):
+    AcademicLevel = apps.get_model('workshops', 'AcademicLevel')
+    L = [
```
Comment (gvwilson) [inline workshops/migrations/0040_auto_20150827_1715.py]: `L`? Really?

File: workshops/migrations/0040_auto_20150827_1715.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        'Don\'t know yet',
+    ]
+    for name in L:
+        AcademicLevel.objects.create(name=name)
+
+
+def add_comp_exp_levels(apps, schema_editor):
+    ComputingExperienceLevel = apps.get_model('workshops',
+                                              'ComputingExperienceLevel')
+    L = [
```
Comment (gvwilson) [inline workshops/migrations/0040_auto_20150827_1715.py]: `L` again?

File: workshops/migrations/0040_auto_20150827_1715.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        'Novice (uses a spreadsheet for data analysis rather than writing '
+            'code)',
```
Comment (gvwilson) [inline workshops/migrations/0040_auto_20150827_1715.py]: I'd put this all on one line even if it is rather long

File: workshops/forms.py (hunk 6 above)
Comment (pbanaszkiewicz) [inline workshops/forms.py]: It is available in SwC form, too. It's just using "Software-Carpentry" in the label / help_text.

File: workshops/migrations/0040_auto_20150827_1715.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        'Intermediate (can write a few lines of code for personal use)',
+        'Proficient (writes multi-page programs which may be shared with '
+            'others)',
```
Comment (gvwilson) [inline workshops/migrations/0040_auto_20150827_1715.py]: As above.

File: workshops/models.py [hunk 11]
```diff
@@ -174,6 +174,105 @@ def save(self, *args, **kwargs):
...
+        max_length=STR_LONG, verbose_name='Family (last) name',
+        blank=False)
+    email = models.CharField(
+        max_length=STR_LONG, verbose_name='Email address', blank=False)
+    affiliation = models.CharField(
+        max_length=STR_LONG,
+        help_text='What university, company, lab, or other organization are '
+        'you affiliated with (if any)?', blank=False)
+    airport_iata = models.CharField(
+        max_length=3, verbose_name='Nearest major airport',
```
Comment (gvwilson) [inline workshops/models.py]: Put `verbose_name` on a line of its own for consistency?

File: workshops/models.py [hunk 12]
```diff
@@ -174,6 +174,105 @@ def save(self, *args, **kwargs):
...
+        ('postdoc', 'Post-doctoral researcher'),
+        ('faculty', 'Faculty'),
+        ('research', 'Research staff (including research programmer)'),
+        ('support', 'Support staff (including technical support)'),
+        ('librarian', 'Librarian/archivist'),
+        ('commerce', 'Commercial software developer '),
+        ('', 'Other (enter below)'),
+    )
+    occupation = models.CharField(
+        max_length=STR_MED, choices=OCCUPATION_CHOICES,
```
Comment (gvwilson) [inline workshops/models.py]: Again, I prefer each field on its own line to make them easier to find.

File: workshops/models.py [hunk 13]
```diff
@@ -174,6 +174,105 @@ def save(self, *args, **kwargs):
...
+        blank=True, default='')
+    github = models.CharField(
+        max_length=STR_LONG, verbose_name='GitHub username',
+        help_text='Please provide your username, not a numeric user ID.',
+        blank=True, default='')
+    twitter = models.CharField(
+        max_length=STR_LONG, verbose_name='Twitter username',
+        help_text='Please, do not put "@" at the beginning.',
+        blank=True, default='')
+    orcid = models.CharField(max_length=STR_LONG, verbose_name='ORCID ID',
```
Comment (gvwilson) [inline workshops/models.py]: Break line after `(` for consistency.

File: workshops/models.py [hunk 14]
```diff
@@ -174,6 +174,105 @@ def save(self, *args, **kwargs):
...
+        verbose_name='Topic and lessons you\'re comfortable teaching',
+        help_text='Please mark ALL that apply.')
+    lessons_other = models.CharField(
+        max_length=255,
+        verbose_name='Other topics/lessons you\'re comfortable teaching',
+        help_text='Please include lesson URLs.',
+        blank=True, default='')
+
+    def save(self, *args, **kwargs):
+        """Try to save nullable char fields as empty strings."""
```
Comment (gvwilson) [inline workshops/models.py]: Why "Try to" ?

File: workshops/models.py [hunk 15]
```diff
@@ -391,6 +490,206 @@ def save(self, *args, **kwargs):
...
+        ('dc', 'Data-Carpentry'),
+    )
+    workshop_type = models.CharField(
+        max_length=STR_MED,
+        choices=WORKSHOP_TYPE_CHOICES,
+        blank=False, default='swc',
+    )
+
+    ATTENDEES_NUMBER_CHOICES = (
+        ('20-40', '20-40 (one room, two instructors)'),
```
Comment (gvwilson) [inline workshops/models.py]: What about less than 20?

File: workshops/models.py [hunk 16]
```diff
@@ -391,6 +490,206 @@ def save(self, *args, **kwargs):
...
+        choices=TRAVEL_REIMBURSEMENT_CHOICES,
+        null=True, blank=True, default=None,
+    )
+    travel_reimbursement_other = models.CharField(
+        max_length=STR_LONG,
+        verbose_name='Other type of reimbursement',
+        blank=True,
+    )
+
+    ADMIN_FEE_PAYMENT_CHOICES = (
```
Comment (gvwilson) [inline workshops/models.py]: Fees have changed - please see http://software-carpentry.org/workshops/request.html for current options.

File: workshops/templates/forms/workshop_request_confirm.html [hunk 17]
```diff
@@ -0,0 +1,9 @@
+{% extends "base_without_navbar.html" %}
+
+{% load crispy_forms_tags %}
+{% load selectable_tags %}
+
+{% block content %}
+<p>Someone will contact you soon.</p>
```
Comment (gvwilson) [inline workshops/templates/forms/workshop_request_confirm.html]: "Thank you for your request - we will be in touch soon."

File: workshops/templates/forms/workshop_request_confirm.html [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+<p>Get back to <a href="https://software-carpentry.org/">software-carpentry.org</a>.</p>
```
Comment (gvwilson) [inline workshops/templates/forms/workshop_request_confirm.html]: Return to `<a href="...">the main web site</a>`.

File: workshops/templates/workshops/eventrequest.html [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+{% block content %}
+{% include "workshops/eventrequest_email_html.html" %}
+
+<div class="clearfix">
+  <p class="edit-object pull-left">
+    {% if perms.workshops.change_eventrequest and perms.workshops.add_event %}
+    <a href="{% url 'eventrequest_accept' object.pk %}" class="btn btn-success">Accept</a>
+    {% else %}
+    <a href="#" class="btn btn-success disabled">Accept</a>
```
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest.html]: Why `href="#"` ?

File: workshops/models.py (hunk 15 above)
Comment (pbanaszkiewicz) [inline workshops/models.py]: This is a default from the old form. Do you want to change it somehow?

File: workshops/templates/workshops/eventrequest_accept.html [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+        <tr><th>Workshop type:</th>
+            <td>{{ object.get_workshop_type_display }}</td></tr>
+        <tr><th>Person:</th>
+            <td>{{ object.name }} &lt;{{ object.email|urlize }}&gt;</td></tr>
+        <tr><th>Affiliation:</th>
+            <td>{{ object.affiliation }}</td></tr>
+        <tr><th>Location:</th>
+            <td>{{ object.location }}</td></tr>
+        <tr><th>Country:</th>
+            <td>{{ object.country.name }} <img src="{{ object.country.flag }}" alt="{{ object.country }}" class="country-flag" /></td></tr>
```
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest_accept.html]: Upper vs lower case for flag names?

File: workshops/templates/workshops/eventrequest_email_html.html [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    <tr><th>Workshop type:</th>
+        <td>{{ object.get_workshop_type_display }}</td></tr>
+    <tr><th>Person:</th>
+        <td>{{ object.name }} &lt;{{ object.email|urlize }}&gt;</td></tr>
+    <tr><th>Affiliation:</th>
+        <td>{{ object.affiliation }}</td></tr>
+    <tr><th>Location:</th>
+        <td>{{ object.location }}</td></tr>
+    <tr><th>Country:</th>
+        <td>{{ object.country.name }} <img src="{{ object.country.flag }}" alt="{{ object.country }}" class="country-flag" /></td></tr>
```
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest_email_html.html]: As above, upper vs lower case for country names.

File: workshops/templates/workshops/eventrequest_email_txt.html [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+Person: {{ object.name }} &lt;{{ object.email|urlize }}&gt;
+Affiliation: {{ object.affiliation }}
+Location: {{ object.location }}
+Country: {{ object.country.name }}
+Alongside of a conference: {{ object.conference|default:"—" }}
+Preferred date: {{ object.preferred_date }}
+Approx. number of attendees: {{ object.get_approx_attendees_display }}
+Attendees' fields: {% for domain in object.attendee_domains.all %}{{ domain }}, {% endfor %}{{ object.attendee_domains_other }}
+Attendees' academic level: {% for level in object.attendee_academic_levels.all %}{{ level }}{% if not forloop.last %}, {% endif %}{% endfor %}
+[SWC] Attendees' level of computing experience: {% for level in object.attendee_computing_levels.all %}{{ level }}{% if not forloop.last %}, {% endif %}{% endfor %}
```
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest_email_txt.html]: Hm - if they've selected the workshop type, can we show/hide the SWC/DC-specific stuff?

File: workshops/templates/workshops/eventrequest_accept.html (hunk 20 above)
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/eventrequest_accept.html]: This actually works on Linux. Could you check if it works on Mac? I think I copied this from the package's sources, so they theoretically should have got it well.

File: workshops/views.py [hunk 23]
```diff
@@ -1381,3 +1387,310 @@ def _failed_to_delete(request, object, protected_objects, back=None):
...
+                'title': 'Thank you for requesting a workshop',
+            }
+            return render(request, 'forms/workshop_request_confirm.html',
+                          context)
+        else:
+            messages.error(request, 'Fix errors below.')
+
+
+class DCEventRequest(SwCEventRequest):
+    form_class = DCEventRequestForm
```
Comment (gvwilson) [inline workshops/views.py]: Why is this just a one-liner when SWC is so complicated??

File: workshops/views.py [hunk 24]
```diff
@@ -1381,3 +1387,310 @@ def _failed_to_delete(request, object, protected_objects, back=None):
     context['refs'].default_factory = None
 
     return render(request, 'workshops/failed_to_delete.html', context)
+
+
+class SwCEventRequest(View):
```
Comment (gvwilson) [inline workshops/views.py]: We use SWC instead of SwC everywhere else.

File: workshops/templates/workshops/eventrequest_email_txt.html (hunk 22 above)
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/eventrequest_email_txt.html]: Yes, we can. Would you like it so?

File: workshops/views.py (hunk 23 above)
Comment (pbanaszkiewicz) [inline workshops/views.py]: Because inheritance.

**Comment by gvwilson** [comment]
This is good progress - please deploy this on a dev version of AMY and let Maneesha drive it around to see what she thinks of the UI.

File: workshops/models.py (hunk 15 above)
Comment (gvwilson) [inline workshops/models.py]: We should allow 1-20 (just for completeness)

File: workshops/templates/workshops/eventrequest_accept.html (hunk 20 above)
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest_accept.html]: If it works, it works :-)

File: workshops/templates/workshops/eventrequest_email_txt.html (hunk 22 above)
Comment (gvwilson) [inline workshops/templates/workshops/eventrequest_email_txt.html]: I *think* so, but check with Maneesha.

**Comment by pbanaszkiewicz** [comment]
>  This is good progress - please deploy this on a dev version of AMY and let Maneesha drive it around to see what she thinks of the UI.

But we don't have a dev server for AMY…

**Comment by gvwilson** [comment]
Didn't we used to?

**Comment by pbanaszkiewicz** [comment]
Maybe we did use Rackspace as a second instance of AMY during the
transition period from Small Orange?

**Comment by gvwilson** [comment]
Hm - @pipitone, wuzzup?

**Comment by pipitone** [comment]
Its true. We don't have a development server. We could easily fire up another rackspace VM for this. 
On Sep 7, 2015 11:52 AM, Greg Wilson <notifications@github.com> wrote:Hm - @pipitone, wuzzup?

File: amy/settings.py (hunk 1 above)
Comment (pbanaszkiewicz) [inline amy/settings.py]: It's Django's default, so I'm only overriding it if the envvar is present.

File: amy/settings.py (hunk 2 above)
Comment (pbanaszkiewicz) [inline amy/settings.py]: You're right.

**Comment by gvwilson** [comment]
Yes please.

File: workshops/templates/workshops/eventrequest.html (hunk 19 above)
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/eventrequest.html]: The links aren't clickable, so there's no point in having a full address in that `href`.

Comment (pbanaszkiewicz) [inline workshops/templates/workshops/eventrequest.html]: The links *that have a class 'disabled'* aren't clickable. *

Comment (gvwilson) [inline workshops/templates/workshops/eventrequest.html]: Ah - OK, then why have the href at all?

Comment (gvwilson) [inline workshops/templates/workshops/eventrequest.html]: Gotcha.

File: workshops/templates/workshops/eventrequest_email_txt.html (hunk 22 above)
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/eventrequest_email_txt.html]: Maneesha prefers to hide irrelevant information.

**Comment by pbanaszkiewicz** [comment]
@gvwilson I'm done with most of the changes you proposed. We only need to setup a dev instance of AMY now.

**Comment by gvwilson** [comment]
Thanks - @pipitone, how quickly can you spin something up?

**Comment by pipitone** [comment]
Piotr, can you take care of creating the AMY dev instance?  I'm unable to until the 12th.

**Comment by pbanaszkiewicz** [comment]
@pipitone sure, but I'm too lazy to do it properly, so it'll run from Django's dev server, is that all right @gvwilson?

**Comment by gvwilson** [comment]
Sure

On 2015-09-08 8:23 PM, Piotr Banaszkiewicz wrote:
>
> @pipitone <https://github.com/pipitone> sure, but I'm too lazy to do 
> it properly, so it'll run from Django's dev server, is that all right 
> @gvwilson <https://github.com/gvwilson>?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/swcarpentry/amy/pull/476#issuecomment-138675777>.
>

-- 
Dr. Greg Wilson    | gvwilson@software-carpentry.org
Software Carpentry | http://software-carpentry.org

**Comment by pbanaszkiewicz** [comment]
@gvwilson I have it partially working. I think the firewall blocks my incoming requests, but it's getting late, so let me deal with it tomorrow.

**Comment by pbanaszkiewicz** [comment]
@gvwilson @pipitone: the dev server is up and running. Could any of you make a domain `amy-dev.software-carpentry.org` to point to `166.78.113.51`?

**Comment by gvwilson** [comment]
Such magic is beyond me - @pipitone, can you cast the appropriate spell?

**Comment by pipitone** [comment]
@pbanaszkiewicz, @gvwilson Done.

**Comment by gvwilson** [comment]
Thanks mate.

**Comment by pbanaszkiewicz** [comment]
Since I didn't receive any further feedback, I'm merging this one.

**Pull request merged**